src/lucky_record/associations.cr

```diff
@@ -39,7 +53,26 @@ module LuckyRecord::Associations
       def preload(preload_query : {{ model }}::BaseQuery)
         add_preload do |records|
           ids = records.map(&.id)
-          {{ assoc_name }} = preload_query.{{ foreign_key }}.in(ids).results.group_by(&.{{ foreign_key }})
+          {% if through %}
+            just_{{ assoc_name }} = preload_query
+              .join_{{ through.id }}
+              .{{ through.id }} do |through_query|
+                through_query.{{ foreign_key.id }}.in(ids)
+              end
+              .preload_{{ through.id }}
+              .distinct
+            {{ assoc_name }} = {} of Int32 => Array(Tag)
+            just_{{ assoc_name }}.each do |item|
+              item.{{ through.id }}.each do |item_through|
+                {{ assoc_name }}[item_through.{{ foreign_key }}] ||= Array({{ model }}).new
+                {{ assoc_name }}[item_through.{{ foreign_key }}] << item
+              end
+            end
```

Comment:
Consider extracting an object for this, though I'm not sure it would work because these are compile-time macros, so I don't know what you could pass to the class to make it work. I'll have to think about this some more